### ISSUE TITLE
Supporting class_weight in train_step ?

### ISSUE THREAD

**Issue by chengjianhong** [tensorflow/tensorflow#43902]
Hi, I have read more about overriding train_step here: https://keras.io/guides/customizing_what_happens_in_fit/. It helps me a lot. But, I still have a problem about support for class_weight in train_step? How to do it? My part code is as follows:

    def train_step(self, data):
        if isinstance(data, tuple):
            x,y,class_weight  = data
        with tf.GradientTape() as tape:

            Flair_encoder_out,T1_encoder_out,T1ce_encoder_out,T2_encoder_out,\
            Flair_decoder_out,T1_decoder_out,T1ce_decoder_out,T2_decoder_out,\
            grade_out = self.MMDR_model(x)

            z_mean_Flair, z_log_var_Flair, z_Flair, common_Flair,spec_Flair = Flair_encoder_out
            z_mean_T1, z_log_var_T1, z_T1, common_T1, spec_T1 =T1_encoder_out
            z_mean_T1ce, z_log_var_T1ce, z_T1ce,common_T1ce, spec_T1ce = T1ce_encoder_out
            z_mean_T2, z_log_var_T2, z_T2,common_T2, spec_T2 = T2_encoder_out

            reconstruction_loss = tf.reduce_mean(
                tf.keras.losses.mean_squared_error(x[0], Flair_decoder_out)
            )+tf.reduce_mean(
                tf.keras.losses.mean_squared_error(x[1], T1_decoder_out)
            )+tf.reduce_mean(
                tf.keras.losses.mean_squared_error(x[2], T1ce_decoder_out)
            )+tf.reduce_mean(
                tf.keras.losses.mean_squared_error(x[3], T2_decoder_out)
            )

            kl_loss = -0.5 * tf.reduce_mean(1 + z_log_var_Flair - tf.square(z_mean_Flair) - tf.exp(z_log_var_Flair))\
                      + (-0.5 * tf.reduce_mean(1 + z_log_var_T1 - tf.square(z_mean_T1) - tf.exp(z_log_var_T1)))\
                      + (-0.5 * tf.reduce_mean(1 + z_log_var_T1ce - tf.square(z_mean_T1ce) - tf.exp(z_log_var_T1ce)))\
                      + (-0.5 * tf.reduce_mean(1 + z_log_var_T2 - tf.square(z_mean_T2) - tf.exp(z_log_var_T2)))

            com_loss = tf.reduce_mean(mean_squared_error(common_Flair, common_T1)) + tf.reduce_mean(mean_squared_error(common_Flair,common_T1ce)) + \
                        tf.reduce_mean(mean_squared_error(common_Flair, common_T2)) + tf.reduce_mean(mean_squared_error(common_T1, common_T1ce)) + \
                        tf.reduce_mean(mean_squared_error(common_T1,common_T2)) + tf.reduce_mean(mean_squared_error(common_T1ce,common_T2))

            spec_loss = tf.reduce_mean(mean_squared_error(spec_Flair, spec_T1)) + tf.reduce_mean(mean_squared_error(spec_Flair, spec_T1ce)) + \
                        tf.reduce_mean(mean_squared_error(spec_Flair, spec_T2)) + tf.reduce_mean(mean_squared_error(spec_T1, spec_T1ce)) + \
                        tf.reduce_mean(mean_squared_error(spec_T1, spec_T2)) + tf.reduce_mean(mean_squared_error(spec_T1ce, spec_T2))
            com_spec_loss = tf.sqrt(com_loss / spec_loss)

            grade_loss = tf.reduce_mean(tf.keras.losses.binary_crossentropy(y,grade_out))

            # grade_loss = self.compiled_loss(
            #     y,
            #     grade_out,
            #     sample_weight=class_weight,
            #     regularization_losses=self.losses,
            # )
            #total_loss = 0.1 * reconstruction_loss + 0.01 * kl_loss + 0.25 * com_spec_loss + grade_loss
            total_loss =  0.1 * reconstruction_loss +  kl_loss + 0.1 * com_spec_loss + grade_loss
            # The loss function is configured in `compile()`.

        # The training happens here.
        grads = tape.gradient(total_loss, self.trainable_variables)

        self.optimizer.apply_gradients(zip(grads, self.trainable_variables))
        print(">>>>>>>>>>>>>>metrics_names:",self.metrics_names)

        # Update metrics (includes the metric that tracks the loss)
        self.compiled_metrics.update_state(y, grade_out)
        return {
            "loss": total_loss,
            "rec_loss": reconstruction_loss,
            "kl_loss": kl_loss,
            "cs_loss": com_spec_loss,
            "grade_loss": grade_loss,
            self.metrics[0].name: self.metrics[0].result(),
            self.metrics[1].name: self.metrics[1].result()
        }

**Comment by Saduf2019** [comment]
@chengjianhong 
Could you please explain the issue faced with "support for class_weight in train_step", if the issue is about how to use you may open a ticket at [stackoverflow](https://stackoverflow.com/questions/tagged/tensorflow) as there is also a larger community that reads questions there.

**Comment by chengjianhong** [comment]
@Saduf2019 I want to use class_weight to solve the class imbalanced problem. But, I use TF2.2 and custom a train_step to achieve multi-task learning. And It does not work to directly add class_weight in model.fit()!

**Comment by jvishnuvardhan** [comment]
@chengjianhong Did you try the example mentioned in https://keras.io/guides/customizing_what_happens_in_fit/ 

Can you please share a simple standalone code to reproduce any error you are facing? Thanks!

**Comment by chengjianhong** [comment]
@jvishnuvardhan The example mentioned in https://keras.io/guides/customizing_what_happens_in_fit/ is about the sample_weight instead of class_weight? My question is how to use class_weight when customizing the train_setp function. My custom train_step function is shown above.

**Comment by nikitamaia** [comment]
Hi @chengjianhong, apologies for the delay here. Please see [this example colab ](https://colab.research.google.com/gist/nikitamaia/e4c0779d0d8d885b76933240986ea697/class_weight_mnist.ipynb) that uses class_weight in train_step to train an MNIST model. Hope this helps!

Closing this issue now. If you need further support help, please post on Stack Overflow.
